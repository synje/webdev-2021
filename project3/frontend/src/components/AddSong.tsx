import { useState } from 'react'
import Form from './AddSongInput'
import { useMutation, gql } from "@apollo/client";
import { InputAddon } from '@chakra-ui/input';
import { Box } from '@chakra-ui/layout';
import { Heading } from '@chakra-ui/react';


/**
 * Mutation-query to VM, adds a new song to the database with the users input
 */
export const ADD_SONG = gql`
mutation AddSong($title:String! $album:String $duration:Float $year_released:Int $artist:ArtistInput ) {
    addSong(songInput:{title:$title album:$album duration:$duration year_released:$year_released artist:$artist }){
        title
    }
}
`

/**
 * Exports AddSong component (everything on the page over the red line) to App.tsx
 */
const AddSong = () => {

    /**
     * States for temporarily storing form-input (onChange)
     */
    let [inputTitle, setInputTitle] = useState("")
    let [inputDuration, setDuration] = useState("")
    let [inputAlbum, setAlbum] = useState("")
    let [inputYear, setYear] = useState("")
    let [inputArtist1, setArtist1] = useState("")
    let [inputArtist2, setArtist2] = useState()

    let FloatDuration = 0
    let IntYear = 0

    /**
     * Mutation to database on submit AddSong-form
     */
    const [addSong, { error, loading, data }] = useMutation(ADD_SONG)
    if (loading) return <p>Submitting...</p>;
    if (error) return <p>Submission error :(( - {error.message}</p>;

    const submit = (variables: any) => {
        addSong(variables)
    }


    return (
        <>
            <Box w="100%" mt="3%" display="flex" alignItems="center" flexDirection="column">
                <Heading display="flex" justifyContent="flex-start" w="90%">Add a song to the collection:</Heading>
                <Box display="flex" justifyContent="center" w="100%" mt="3%" mb="3%" borderBottom="3px solid darkred">
                    {/**
                     * Form is a component from the file AddSongInput.tsx.
                     * Form is the input-fields and the button to add song.
                     * 
                     * Stores values from input-fields onChange, submits values to mutation-query
                     */}
                    <Form
                        titleInput={(e: any) => inputTitle = (e.target.value)}
                        durationInput={(e: any) => inputDuration = (e.target.value)}
                        albumInput={(e: any) => inputAlbum = (e.target.value)}
                        yearInput={(e: any) => inputYear = (e.target.value)}
                        artistInput1={(e: any) => inputArtist1 = e.target.value}
                        artistInput2={(e: any) => inputArtist2 = e.target.value}

                        handleClick={(e: any) => {
                            FloatDuration = parseFloat(inputDuration)
                            IntYear = parseInt(inputYear)
                            submit({ variables: { title: inputTitle, duration: FloatDuration, album: inputAlbum, year_released: IntYear, artist: { name1: inputArtist1, name2: inputArtist2 } } })
                        }}
                    />
                </Box>
            </Box>
        </>
    )
}

export default AddSong;