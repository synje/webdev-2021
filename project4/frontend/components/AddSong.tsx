import { gql, useMutation } from "@apollo/client";
import React from "react";
import { Controller, useForm } from "react-hook-form";
import { View, Text, StyleSheet, Platform, ScrollView, Dimensions } from "react-native";
import { Button, Headline, TextInput } from "react-native-paper";

/**
 * Mutation-query to VM, adds a new song to the database with the users input
 */
export const ADD_SONG = gql`
 mutation AddSong($title:String! $album:String $duration:Float $year_released:Int $artist:ArtistInput ) {
     addSong(songInput:{title:$title album:$album duration:$duration year_released:$year_released artist:$artist }){
         title
     }
 }
 `

/**
 * Component for adding song to the database
 * @returns JSX elements for data input
 */
function AddSong() {

    const [addSong, { error, loading, data }] = useMutation(ADD_SONG)
    const { control, handleSubmit, formState: { errors }, reset } = useForm();

    const resetValues = {
        title: '',
        duration: '',
        album: '',
        year: '',
        artist1: '',
        artist2: ''
    }

    /**
     * Takes input data as string and replaces , with . to make the duration format correct on web
     * @param input 
     * @returns input with .
     */
    function makeStringToFloat(input: string) {
        return parseFloat(input.replace(',', '.'))
    }

    /**
     * Feedback to user that a song was added successfully
     * @param data 
     */
    function songAddedSuccessful(data: any) {
        alert(`Song "${data.title}" was added!`);
    }

    /**
     * Checks to see of artist2 input is empty to format the list element correctly. If artist is empty sets the value as undefined, otherwise artist2 is returned as it is
     * @param artist2 
     * @returns artist2 or undefined
     */
    function checkArtist2(artist2: string) {
        if (artist2 !== "") {
            return artist2
        }
        return undefined

    }

    /**
     * Submits the input data as a new song to the database
     * @param data 
     * @returns alert if an error occurs, or else it will smoothly reload the form
     */
    const onSubmit = (inputData: any) => {
        const inputDataVar = { variables: { title: inputData.title.toString(), duration: makeStringToFloat(inputData.duration), album: inputData.album.toString(), year_released: parseInt(inputData.year), artist: { name1: inputData.artist1, name2: checkArtist2(inputData.artist2) } } }
        addSong(inputDataVar)
        if (loading) return <Text style={{ marginTop: 10 }}>Submitting...</Text>;
        if (error) return <Text style={{ marginTop: 10 }}>Submission error: {error.message}</Text>;
        if (data !== undefined) {
            songAddedSuccessful(inputData)
        }
        else {
            alert('An error has occured. An error occured while loading. You might not be connected to the NTNU network or your VPN is off.');
        }
        reset(resetValues)
    }

    //Source/Inspiration: https://stackoverflow.com/questions/69370034/how-to-input-only-number-in-react-hook-form
    /**
     * Makes the inputfield for year only accept numbers 
     * @param value 
     * @returns string input as "numbers"
     */
    const allowOnlyNumber = (value: string) => {
        return value.replace(/[^0-9]/g, '')
    }


    return (
        <>
            <View style={styles.body}>
                <ScrollView style={styles.scroll}>
                    <View style={styles.container}>
                        <View>
                            <Headline style={styles.headline}>ADD SONG</Headline>
                        </View>
                        <Controller
                            control={control}
                            rules={{
                                required: true,
                            }}
                            render={({ field: { onChange, onBlur, value } }) => (
                                <TextInput
                                    clearButtonMode="always"
                                    value={value}
                                    onBlur={onBlur}
                                    onChangeText={value => onChange(value)}
                                    label="Song title *"
                                    style={styles.input}
                                    accessibilityLabel='Song title *'
                                    placeholder="Take Me To Church"
                                    placeholderTextColor={'#79b298'}
                                    autoComplete="off"
                                />
                            )}
                            name="title"
                            defaultValue=""
                        />
                        {errors.title && <Text>Field is required.</Text>}


                        <Controller
                            control={control}
                            rules={{
                                required: true,
                                value: Number,
                                pattern: /^[0-9.,]*$/
                            }}
                            render={({ field: { onChange, onBlur, value } }) => (
                                <TextInput
                                    onBlur={onBlur}
                                    onChangeText={value => onChange(value)}
                                    value={value}
                                    keyboardType="numeric"
                                    label="Duration *"
                                    placeholder="4.01"
                                    placeholderTextColor={'#79b298'}
                                    style={styles.input}
                                    accessibilityLabel='Duration *'
                                    autoComplete="off"
                                />
                            )}
                            name="duration"
                            defaultValue=""
                        />
                        {errors.duration && errors.duration.type === 'required' && <Text>Field is required.</Text>}
                        {errors.duration && errors.duration.type === 'pattern' && <Text>Field must be numbers.</Text>}

                        <Controller
                            control={control}
                            rules={{
                                required: true,
                            }}
                            render={({ field: { onChange, onBlur, value } }) => (
                                <TextInput
                                    value={value}
                                    onBlur={onBlur}
                                    onChangeText={value => onChange(value)}
                                    label="Album *"
                                    style={styles.input}
                                    autoComplete="off"
                                    accessibilityLabel='Album *'
                                    placeholder="Hozier"
                                    placeholderTextColor={'#79b298'}
                                />
                            )}
                            name="album"
                            defaultValue=""
                        />
                        {errors.album && <Text>Field is required.</Text>}

                        <Controller
                            control={control}
                            rules={{
                                required: true,
                                maxLength: 4
                            }}
                            render={({ field: { onChange, onBlur, value } }) => (
                                <TextInput
                                    onBlur={onBlur}
                                    onChangeText={(text) => onChange(allowOnlyNumber(text))}
                                    value={value}
                                    keyboardType="numeric"
                                    label="Year Released *"
                                    style={styles.input}
                                    autoComplete="off"
                                    accessibilityLabel='Year released *'
                                    placeholder="2014"
                                    placeholderTextColor={'#79b298'}
                                />
                            )}
                            name="year"
                            defaultValue=""
                        />
                        {errors.year && errors.year.type === 'required' && <Text>Field is required.</Text>}
                        {errors.year && errors.year.type === 'maxLength' && <Text>Field can not contain more than 4 numbers.</Text>}

                        <Controller
                            control={control}
                            rules={{
                                required: true,
                            }}
                            render={({ field: { onChange, onBlur, value } }) => (
                                <TextInput
                                    value={value}
                                    onBlur={onBlur}
                                    onChangeText={value => onChange(value)}
                                    label="Artist 1 *"
                                    style={styles.input}
                                    autoComplete="off"
                                    accessibilityLabel='Artist 1 *'
                                    placeholder="Hozier"
                                    placeholderTextColor={'#79b298'}
                                />
                            )}
                            name="artist1"
                            defaultValue=""
                        />
                        {errors.artist1 && <Text>Field is required.</Text>}

                        <Controller
                            control={control}
                            rules={{
                                required: false,
                            }}
                            render={({ field: { onChange, onBlur, value } }) => (
                                <TextInput
                                    onBlur={onBlur}
                                    onChangeText={onChange}
                                    value={value}
                                    label="Artist 2"
                                    style={styles.input}
                                    autoComplete="off"
                                    accessibilityLabel='Artist 2'
                                />
                            )}
                            name="artist2"
                            defaultValue=""
                        />

                        <Button mode="contained" onPress={handleSubmit(onSubmit)} style={styles.button}>SUBMIT</Button>
                    </View>
                </ScrollView>
            </View>
        </>
    )
}
export default AddSong



const screenHeight = Platform.OS === "web" ? Dimensions.get('screen').height + 20 : Dimensions.get('screen').height
const windowHeight = Platform.OS === "web" ? Dimensions.get('window').height + 20 : Dimensions.get('window').height


const styles = StyleSheet.create({
    body: {
        width: '100%',
        height: windowHeight - 150,
        marginTop: 0,
    },
    container: {
        width: '100%',
        maxHeight: '100%',
        alignItems: 'center',
    },
    input: {
        width: "70%",
        maxWidth: 900,
        marginTop: Platform.OS === "web" ? 20 : 10,
        shadowColor: '#0c1618',
        shadowOffset: { width: 3, height: 3 },
        shadowOpacity: 0.3,
        shadowRadius: 5,
        borderWidth: 1,
        borderRadius: 5,
        borderColor: "#004643",
        backgroundColor: '#F0FAF1',
    },
    button: {
        width: 250,
        height: 50,
        justifyContent: 'center',
        marginTop: Platform.OS === "web" ? 20 : 10,
        marginBottom: Platform.OS === "web" ? 0 : screenHeight / 3 + 10,
    },
    headline: {
        fontSize: 30,
        fontWeight: 'bold',
        marginBottom: 25,
        marginTop: 20,
    },
    scroll: {
        height: Platform.OS === "web" ? screenHeight / 1.5 : screenHeight - 150,
        marginTop: Platform.OS === "web" ? 0 : 0,
        width: '100%',
        paddingBottom: Platform.OS === "web" ? 20 : 50,
    },

});