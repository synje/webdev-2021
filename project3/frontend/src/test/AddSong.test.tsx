import { useMutation } from "@apollo/client";
import { MockedProvider } from "@apollo/react-testing";
import { fireEvent, getByLabelText, getByText, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Provider } from "react-redux";
import AddSong, { ADD_SONG } from "../components/AddSong";
export { ADD_SONG } from '../components/AddSong';

import { ApolloClient, InMemoryCache, ApolloProvider, } from "@apollo/client";
import TestRenderer, { ReactTestRendererJSON } from 'react-test-renderer';
import renderer from 'react-test-renderer'
import { Button } from "@chakra-ui/button";
import { GraphQLError } from "graphql/error";


const mockAddSong = {
    request: {
        query: ADD_SONG,
        variables: {
            title: "RED-SUN (021)",
            duration: 3.09,
            album: "THE PARK IN THE NIGHT part three",
            year_released: 2019,
            aritst: {
                name1: 'GWSN',
            },
        },
    },
    result: {
        data: {
            filterSongs: {
                _id: "1234",
                title: "Sang",
                duration: 1.23,
                album: "album",
                year_released: 2021,
                artist: {
                    name1: "artist",
                    name2: undefined
                }
            }
        }
    }
};



test('Snapshot addSong', () => {
    const tree = renderer.create(<MockedProvider mocks={[mockAddSong as any]}>
        <AddSong />
    </MockedProvider>)
    expect(tree).toMatchSnapshot()

})
it('Test render doesnt result in error', async () => {
    const { getByText } = render(
        <MockedProvider mocks={[mockAddSong as any]}>
            <AddSong />
        </MockedProvider>
    );

    expect(getByText(/Song title/i)).toBeInTheDocument();

    //https://www.apollographql.com/docs/react/v2/development-testing/testing/

});
