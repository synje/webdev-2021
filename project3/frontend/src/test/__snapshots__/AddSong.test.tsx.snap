// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot addSong 1`] = `
<div
  className="css-1is5g6e"
>
  <div
    className="css-emzx1y"
  >
    <div
      className="chakra-form-control css-uewl2b"
      role="group"
    >
      <label
        className="chakra-form__label css-uvho5g"
        htmlFor="songtitleInput"
        id="field-4015174717-1-label"
        style={
          Object {
            "fontSize": "120%",
          }
        }
      >
        Song title:
        <span
          aria-hidden={true}
          className="chakra-form__required-indicator css-0"
          role="presentation"
        >
          *
        </span>
      </label>
      <input
        aria-required={true}
        className="chakra-input addSongInput css-0"
        data-cy="title"
        disabled={false}
        id="field-4015174717-1"
        name="songtitle"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Title"
        readOnly={false}
        required={true}
        type="text"
      />
    </div>
    <div
      className="chakra-form-control css-uewl2b"
      role="group"
    >
      <label
        className="chakra-form__label css-uvho5g"
        htmlFor="durationInput"
        id="field-4015174717-2-label"
        style={
          Object {
            "fontSize": "120%",
          }
        }
      >
        Duration:
        <span
          aria-hidden={true}
          className="chakra-form__required-indicator css-0"
          role="presentation"
        >
          *
        </span>
      </label>
      <input
        aria-required={true}
        className="chakra-input addSongInput css-0"
        data-cy="duration"
        disabled={false}
        id="field-4015174717-2"
        name="durationInput"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="2.45"
        readOnly={false}
        required={true}
        type="number"
      />
      <div
        className="chakra-form__helper-text css-1q3xmzu"
        id="field-4015174717-2-helptext"
      >
        Song length in minutes comma seconds
      </div>
    </div>
    <div
      className="chakra-form-control css-uewl2b"
      role="group"
    >
      <label
        className="chakra-form__label css-uvho5g"
        htmlFor="albumInput"
        id="field-4015174717-3-label"
        style={
          Object {
            "fontSize": "120%",
          }
        }
      >
        Album:
        <span
          aria-hidden={true}
          className="chakra-form__required-indicator css-0"
          role="presentation"
        >
          *
        </span>
      </label>
      <input
        aria-required={true}
        className="chakra-input addSongInput css-0"
        data-cy="album"
        disabled={false}
        id="field-4015174717-3"
        name="albumInput"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Album"
        readOnly={false}
        required={true}
        type="text"
      />
    </div>
    <div
      className="chakra-form-control css-uewl2b"
      role="group"
    >
      <label
        className="chakra-form__label css-uvho5g"
        htmlFor="yearInput"
        id="field-4015174717-4-label"
        style={
          Object {
            "fontSize": "120%",
          }
        }
      >
        Year released:
        <span
          aria-hidden={true}
          className="chakra-form__required-indicator css-0"
          role="presentation"
        >
          *
        </span>
      </label>
      <input
        aria-required={true}
        className="chakra-input addSongInput css-0"
        data-cy="year"
        disabled={false}
        id="field-4015174717-4"
        name="yearInput"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="1999"
        readOnly={false}
        required={true}
        type="number"
      />
    </div>
    <div
      className="chakra-form-control css-uewl2b"
      role="group"
    >
      <label
        className="chakra-form__label css-uvho5g"
        htmlFor="artist1Input"
        id="field-4015174717-5-label"
        style={
          Object {
            "fontSize": "120%",
          }
        }
      >
        Artist 1:
        <span
          aria-hidden={true}
          className="chakra-form__required-indicator css-0"
          role="presentation"
        >
          *
        </span>
      </label>
      <input
        aria-required={true}
        className="chakra-input addSongInput css-0"
        data-cy="artist1"
        disabled={false}
        id="field-4015174717-5"
        name="artist1Input"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Name Johnson"
        readOnly={false}
        required={true}
        type="String"
      />
    </div>
    <div
      className="chakra-form-control css-d2znx6"
      role="group"
    >
      <label
        className="chakra-form__label css-uvho5g"
        htmlFor="artist2Input"
        id="field-4015174717-6-label"
        style={
          Object {
            "fontSize": "120%",
          }
        }
      >
        Artist 2:
      </label>
      <input
        className="chakra-input addSongInput css-0"
        data-cy="artist2"
        disabled={false}
        id="field-4015174717-6"
        name="artist2Input"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="John Nameson"
        readOnly={false}
        required={false}
        type="String"
      />
    </div>
    <div
      className="css-1l4w6pd"
    >
      <button
        className="chakra-button submitbtn css-hek6jo"
        id="AddSongButton"
        onClick={[Function]}
        type="submit"
      >
        ADD SONG
      </button>
    </div>
  </div>
</div>
`;
